NumLib/ODESolver/ConvergenceCriterionResidual.cpp

```diff
@@ -43,26 +43,36 @@ void ConvergenceCriterionResidual::checkResidual(const GlobalVector& residual)
 {
     auto norm_res = MathLib::LinAlg::norm(residual, _norm_type);
 
-    if (_is_first_iteration) {
+    if (_is_first_iteration)
+    {
         INFO("Convergence criterion: |r0|=%.4e", norm_res);
-        _residual_norm_0 = norm_res;
-    } else {
+        _residual_norm_0 =
+            (norm_res < std::numeric_limits<double>::min()) ? 1.0 : norm_res;
```

Comment:
Changed, and the default value is dropped.

The check here tells that _residual_norm_0 is zero is zero, it takes the current norm.